libmusig2.c

```diff
@@ -1,282 +1,359 @@
 #include "libmusig2.h"
-#include <sodium.h>
-#include <stdio.h>
-#include <string.h>
-#include <assert.h>
 
-// Generate key-pair.
-void keypair_gen(unsigned char *sk, unsigned char *pk) {
 
-    crypto_core_ristretto255_scalar_random(sk);
-    crypto_scalarmult_ristretto255_base(pk, sk);
+void INIT_Signer(secp256k1_context* ctx, SIGNER signer){
+
+    size_t point_ser_len = PK_SIZE;
+    secp256k1_pubkey* pubkey = (secp256k1_pubkey*)malloc(sizeof(secp256k1_pubkey));
+
+	signer->sk = (unsigned char*)malloc(SK_SIZE);
+	signer->PK = (unsigned char*)malloc(PK_SIZE);
+
+	while (1){
+        /* Generate a random secret key. */
+		if (!fill_random(signer->sk, SK_SIZE)){
+            printf("Failed to generate secret key!\n");
+            return;
+		}
+        /* Generate the corresponding public key and serialize it in signer->PK. */
+		if (secp256k1_ec_pubkey_create(ctx, pubkey, signer->sk) &&
+                secp256k1_ec_pubkey_serialize(ctx, signer->PK, &point_ser_len, pubkey, SECP256K1_EC_COMPRESSED))
+            break;
+	}
 }
 
-// Function that generates randomness and its corresponding commitment, for a single signature.
-void commit(unsigned char *commitment, unsigned char *randomness) {
-    return batch_commit(commitment, randomness, 1);
+void SIGN_FirstRound(const secp256k1_context* ctx, SIGNER signer){
+
+    int j;
+    size_t point_ser_len = PK_SIZE;
+    secp256k1_pubkey* pubkey = (secp256k1_pubkey*)malloc(sizeof(secp256k1_pubkey));
+
+    signer->r = (unsigned char**)malloc(SK_SIZE*V);
+    signer->R = (unsigned char**)malloc(PK_SIZE*V);
+
+    for(j=0; j<V;j++){
+		signer->r[j] = (unsigned char*)malloc(SK_SIZE);
+        signer->R[j] = (unsigned char*)malloc(PK_SIZE);
+
+		while (1) {
+            /* Generate a random r[j]. */
+			if (!fill_random(signer->r[j], SK_SIZE)) {
+                printf("Failed to generate randomness!\n");
+                return;
+			}
+            /* Generate the corresponding nonce R and serialize it in R[j]. */
+            if (secp256k1_ec_pubkey_create(ctx, pubkey, signer->r[j]) &&
+                    secp256k1_ec_pubkey_serialize(ctx, signer->R[j], &point_ser_len, pubkey, SECP256K1_EC_COMPRESSED))
+                break;
+		}
+    }
 }
 
-// Function that generates randomness and their corresponding commitments, for `batch_size` signatures.
-void batch_commit(unsigned char *commitment, unsigned char *randomness, unsigned long batch_size) {
-    for (int j = 0; j < batch_size; j ++){
-        for (int i = 0; i < NR_V; i++) {
-            crypto_core_ristretto255_scalar_random(randomness + j * NR_V * crypto_core_ristretto255_SCALARBYTES + i * crypto_core_ristretto255_SCALARBYTES);
+void COLLECT_Signers(AGG_ROUND1 aggRound1, const unsigned char* pk, unsigned char** R, int i)
```

Comment:
seems like this function would be a good place to deserialise the points, and check their validity.